### PR TITLE
test: edge cases for server action

### PR DESCRIPTION
**Pull request by himself65**
I was trying to make vercel-ai + waku, but found some bugs, let me see if I can reproduce here

### CODE REVIEW

File: e2e/fixtures/rsc-basic/modules/ai/src/server.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+async function innerAction({ action, options }, state, ...args) {
+  'use server';
+  const result = await action(...args);
+  // eslint-disable-next-line no-undef
+  console.log('wrapped action', result);
+  return result;
+}
+
+function wrapAction(action, options) {
+  return innerAction.bind(null, { action, options });
```
Comment (himself65) [inline e2e/fixtures/rsc-basic/modules/ai/src/server.js:14]: Wrap action will cause error for waku but okay for nextjs

@dai-shi

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 2]
```diff
@@ -188,7 +198,46 @@ export async function renderRsc(
     ) {
       // XXX This doesn't support streaming unlike busboy
       const formData = parseFormData(bodyStr, contentType);
-      args = await decodeReply(formData);
```
Comment (himself65) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:191]: There are something wrong, formData could include server references which need webpack runtime @dai-shi

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 3]
```diff
@@ -188,7 +198,46 @@ export async function renderRsc(
...
+      const moduleMap = new Proxy({} as Record<string, ImportManifestEntry>, {
+        get(_target, rsfId: string): ImportManifestEntry {
+          const [fileId, name] = rsfId.split('#') as [string, string];
+          // fixme: race condition, server actions are not initialized in the first time
```
Comment (himself65) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:204]: As said, I think we should initialize all server actions before running RSC otherwise it will crash for the first time

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 4]
```diff
@@ -188,7 +198,46 @@ export async function renderRsc(
...
+              );
+            } else {
+              moduleLoading.set(
+                fileId,
+                opts
+                  .loadServerFile(filePathToFileURL(fileId))
+                  .then((m: any) => {
+                    moduleCache.set(
+                      fileId,
+                      Object.fromEntries(m['__waku_serverActions']),
```
Comment (himself65) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:227]: Looks ugly

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 5]
```diff
@@ -188,7 +198,46 @@ export async function renderRsc(
...
+                    );
+                  }),
+              );
+            }
+          }
+          return {
+            id: fileId,
+            chunks: [],
```
Comment (himself65) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:235]: `chunks: [fileId]` is not necessary after reading the source code fo next.js

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 3 above)
Comment (himself65) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:204]: for next.js, they have a global server action map in the runtime

https://github.com/vercel/next.js/blob/86dbcbea431f2fd85d94a72a9c2d35781154a5d7/packages/next/src/server/app-render/encryption-utils.ts#L138-L158

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 2 above)
Comment (himself65) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:191]: https://github.com/facebook/react/blob/ddcecbbebf6af3fa32a323c0591dad1f63587aeb/packages/react-server/src/ReactFlightReplyServer.js#L927-L941

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts [hunk 6]
```diff
@@ -5,10 +5,25 @@ import { EXTENSIONS } from '../config.js';
 import { extname } from '../utils/path.js';
 import { parseOpts } from '../utils/swc.js';
 
-const collectExportNames = (mod: swc.Module) => {
+const collectExportNames = (
```
Comment (himself65) [inline packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:8]: Here are other issues:

1. non export sever action could cause runtime error
2. nested server action could cause runtime error

```ts
const actionsMap = {
  foo: () => {
     "use server"
// ...
  } 
}
```

Those are exist in vercel/ai/rsc source so we need to support them

**Review by himself65 — COMMENTED** [review]
This PR little mess up, but should separate into three parts: 

1. Test some RSC edge cases
2. RSC transform plugin
3. RSC module map resolution

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts (hunk 6 above)
Comment (dai-shi) [reply packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:8]: Okay, I missed such use cases.
This pattern, inline "use server", is only allowed in the server files, right?

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts [hunk 7]
```diff
@@ -223,40 +238,46 @@ const transformServer = (
...
 import { registerClientReference } from 'react-server-dom-webpack/server';
 `;
-    for (const name of exportNames) {
+    for (const name of [...exports, ...internal]) {
       newCode += `
 export ${name === 'default' ? name : `const ${name} =`} registerClientReference(() => { throw new Error('It is not possible to invoke a client function from the server: ${getClientId(id)}#${name}'); }, '${getClientId(id)}', '${name}');
 `;
     }
     return newCode;
-  } else if (hasUseServer) {
```
Comment (dai-shi) [inline packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:236]: I'm not sure if we can remove this branch. It's for another use case unless I'm mistaken.

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts [hunk 8]
```diff
@@ -223,40 +238,46 @@ const transformServer = (
...
+      if (name === 'default') {
+        continue;
+      }
       newCode += `
 if (typeof ${name} === 'function') {
   registerServerReference(${name}, '${getServerId(id)}', '${name}');
 }
 `;
+      if (internal.has(name)) {
+        newCode += `export { ${name} };\n`;
```
Comment (dai-shi) [inline packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:276]: This doesn't feel correct. Possible name collisions.

File: packages/waku/src/lib/renderers/html-renderer.ts [hunk 9]
```diff
@@ -260,17 +260,7 @@ export const renderHtml = async (
...
-      Record<
-        string,
-        {
-          id: string;
-          chunks: string[];
-          name: string;
-        }
-      >
-    >,
+    {} as Record<string, Record<string, ImportManifestEntry>>,
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/html-renderer.ts:263]: Nice, but maybe we should have a separate PR.

File: packages/waku/src/lib/renderers/rsc-renderer.ts [hunk 10]
```diff
@@ -12,6 +12,16 @@ import { parseFormData } from '../utils/form.js';
 import { streamToString } from '../utils/stream.js';
 import { decodeActionId } from '../renderers/utils.js';
 
+// HACK for react-server-dom-webpack without webpack
```
Comment (dai-shi) [inline packages/waku/src/lib/renderers/rsc-renderer.ts:15]: It seems like we should finish up #707 before this PR.

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 2 above)
Comment (dai-shi) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:191]: Alright. It's reasonable that it requires the second argument `moduleMap`.

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 3 above)
Comment (dai-shi) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:204]: I still need to understand the issue, but it feels like we want lazy evaluation.

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 4 above)
Comment (dai-shi) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:227]: Yeah... I hope we can improve it.

File: packages/waku/src/lib/renderers/rsc-renderer.ts (hunk 5 above)
Comment (dai-shi) [reply packages/waku/src/lib/renderers/rsc-renderer.ts:235]: Are you sure? I mean waku's approach might be different from nextjs's one.

**Review by dai-shi — COMMENTED** [review]
This looks like a good improvement.
If possible, we would like to split into multiple PRs and merge gradually.
(And, we definitely need unit tests for transformation. 😓 )

**Comment by himself65** [comment]
> Are you sure? I mean waku's approach might be different from nextjs's one.

It depends on what's the real behavior from webpack, I need double check that

File: packages/waku/src/lib/renderers/html-renderer.ts (hunk 9 above)
Comment (dai-shi) [reply packages/waku/src/lib/renderers/html-renderer.ts:263]: https://github.com/dai-shi/waku/pull/769

File: packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts (hunk 6 above)
Comment (dai-shi) [reply packages/waku/src/lib/plugins/vite-plugin-rsc-transform.ts:8]: Hm, I made a test case #771, but it looks like it's transforming well.
Can you give me a failing case?

This is also tried:
```js
const actionsMap = createActions({
  foo: () => {
     "use server"
     // ...
  } 
})
```

**Comment by dai-shi** [comment]
I reverted the src code, but why do all e2e tests fail... should only a new one fail??

**Comment by himself65** [comment]
no it's all basic rsc test, I think something is still wrong

**Comment by himself65** [comment]
For this input:

```tsx
const AI = createAI({
	actions: {
		foo: async function() {
			"use server";
			return 0;
		}
	}
});
export { AI }
```

should emit something like

```tsx
const AI = createAI({actions: {foo: __waku__server__1 });
export { AI };
export const __waku__server__1 = registerServerReference(async function() {
	'use server';
	return 0;
}, 'file_id', '1');
```

**Comment by himself65** [comment]
working on `decodeReply`

**Comment by himself65** [comment]
PR is to complex to rebease/merge, create a new one https://github.com/dai-shi/waku/pull/785
